File: src/marvin/schemas/platform/blueprints.py
"""Blueprint schemas — declarative descriptions of workspace structure.

A *blueprint* says "here is a collection / entry type / scheduled task a workspace could have",
without creating anything. Core ships a catalog of them; integration providers contribute their own
(a provider may only ever *declare* — the SDK forbids it touching the database). One schema, three
consumers: the catalog a user browses, what an integration brings on install, and what the agent
offers to create.

Applying a blueprint is an upsert by slug that **creates what is missing and never overwrites what
exists** — a workspace may have customised its copy, and a blueprint is a starting point, not a
template the workspace is held to.
"""

from typing import Annotated, Literal

from pydantic import Field, StringConstraints, field_validator

from marvin.schemas._marvin import _MarvinModel

BlueprintKind = Literal["collection", "entry_type", "scheduled_task"]

#: `requires` entries look like "entry_type:bench-note" — a blueprint that references content the
#: workspace does not have is surfaced as unavailable rather than applied into a broken state.
REQUIREMENT_KINDS = ("entry_type", "collection")

CORE_SOURCE = "core"


class Blueprint(_MarvinModel):
    """One thing a workspace could create, described rather than created."""

    kind: BlueprintKind
    """What this blueprint builds."""

    slug: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]
    """Slug the created object gets, and the key the upsert dedupes on."""

    name: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]
    """Human label, shown in the catalog."""

    description: str = ""
    """One line on what it is for — this is the discoverability, so it earns its place."""

    category: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)] = "General"
    """Groups the catalog. Core uses editorial/workflow-style categories; a provider's blueprints
    are filed under that provider so the catalog stays legible as providers multiply."""

    source: Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)] = CORE_SOURCE
    """`core`, or the slug of the provider that contributed it."""

    requires: list[str] = Field(default_factory=list)
    """Prerequisites as `kind:slug`, e.g. `entry_type:bench-note`. Checked before offering."""

    payload: dict = Field(default_factory=dict)
    """The body handed to the creating repository — a CollectionCreate / EntryTypes /
    ScheduledTaskCreate shape, minus the workspace scoping the core adds."""

    @field_validator("requires")
    @classmethod
    def _validate_requires(cls, value: list[str]) -> list[str]:
        for req in value:
            kind, _, slug = req.partition(":")
            if kind not in REQUIREMENT_KINDS or not slug:
                raise ValueError(f"requirement must be '<{'|'.join(REQUIREMENT_KINDS)}>:<slug>', got {req!r}")
        return value

    @field_validator("payload")
    @classmethod
    def _payload_carries_no_scope(cls, value: dict) -> dict:
        # Scoping is the core's job. A blueprint that carried its own group_id could be applied into
        # the wrong workspace, so refuse the shape outright rather than silently dropping the key.
        for forbidden in ("group_id", "id"):
            if forbidden in value:
                raise ValueError(f"payload must not set {forbidden!r} — the workspace is supplied when the blueprint is applied")
        return value


class BlueprintRead(Blueprint):
    """A catalog entry as the API returns it, with what this workspace can do about it."""

    available: bool = True
    """False when `requires` is unmet — shown, but not applicable."""

    missing_requirements: list[str] = Field(default_factory=list)
    """The unmet entries from `requires`, so the UI can say what is needed."""

    applied: bool = False
    """True when this workspace already has something with that slug — applying again is a no-op."""


class BlueprintApplyResult(_MarvinModel):
    """What applying one blueprint did."""

    slug: str
    kind: BlueprintKind
    created: bool
    """False when it already existed; the existing object is left exactly as it was."""

    detail: str = ""
    """Why nothing was created, when nothing was."""

File: frontend/src/lib/api/blueprints.ts
/**
 * Blueprints — the catalog of structure a workspace could have, and applying it.
 *
 * A blueprint describes a collection, entry type or scheduled task without creating anything.
 * Core ships a catalog of examples; installed integration providers contribute the content their
 * actions depend on. Applying creates what is missing and never overwrites what exists, so a
 * repeat apply is a no-op that says so.
 *
 * There is deliberately no blueprint gallery. The catalog is consumed where it is useful: as
 * examples offered while you create a collection or entry type, as the content an integration
 * brings on install, and as the worked examples that teach an agent the rule vocabulary well
 * enough to compose rules of its own.
 *
 * Browser calls go through the same-origin proxy (fetchApi with no token); SSR passes the cookie token.
 */

import { fetchApi } from "./client";

export type BlueprintKind = "collection" | "entry_type" | "scheduled_task";

/** `entry_type`/`collection` render as a picker fed by the workspace's own content. */
export type ParameterKind = "entry_type" | "collection" | "text" | "number";

export interface BlueprintParameter {
  key: string;
  label: string;
  kind: ParameterKind;
  required: boolean;
  default: string | null;
  help: string;
}

export interface Blueprint {
  kind: BlueprintKind;
  slug: string;
  name: string;
  description: string;
  category: string;
  /** `core`, or the slug of the provider that contributed it. */
  source: string;
  requires: string[];
  parameters: BlueprintParameter[];
  payload: Record<string, unknown>;
  /** False when `requires` is unmet — shown, but not applicable. */
  available: boolean;
  missingRequirements: string[];
  /** Already in this workspace. Always false for a parameterised blueprint: its slug
   *  is not known until the parameters are. */
  applied: boolean;
}

export interface BlueprintApplyResult {
  slug: string;
  kind: BlueprintKind;
  created: boolean;
  detail: string;
  name: string;
}

export interface BlueprintFilters {
  kind?: BlueprintKind;
  category?: string;
  source?: string;
}

function query(filters: BlueprintFilters = {}): string {
  const params = new URLSearchParams();
  for (const [key, value] of Object.entries(filters)) {
    if (value) params.set(key, value);
  }
  const qs = params.toString();
  return qs ? `?${qs}` : "";
}

/** The catalog, annotated with what this workspace can do about each entry. */
export async function listBlueprints(filters: BlueprintFilters = {}, authToken?: string): Promise<Blueprint[]> {
  return fetchApi<Blueprint[]>(`/api/groups/blueprints${query(filters)}`, {}, authToken);
}

/** Category names in catalog order — core's first, then each provider's. */
export async function listBlueprintCategories(authToken?: string): Promise<string[]> {
  return fetchApi<string[]>("/api/groups/blueprints/categories", {}, authToken);
}

/** One blueprint including its payload — what a prefilled editor reads. */
export async function getBlueprint(slug: string, source?: string, authToken?: string): Promise<Blueprint> {
  return fetchApi<Blueprint>(`/api/groups/blueprints/${encodeURIComponent(slug)}${query({ source })}`, {}, authToken);
}

/** Create this blueprint's object, unless something already has its slug. */
export async function applyBlueprint(
  slug: string,
  params?: Record<string, string>,
  source?: string,
  authToken?: string,
): Promise<BlueprintApplyResult> {
  return fetchApi<BlueprintApplyResult>(
    `/api/groups/blueprints/${encodeURIComponent(slug)}/apply${query({ source })}`,
    { method: "POST", body: JSON.stringify(params ?? null), headers: { "Content-Type": "application/json" } },
    authToken,
  );
}

/**
 * Apply several — what the "apply this integration's content" button sends. Entry types are
 * created before the collections and tasks that reference them, whatever order they arrive in.
 * `params` is keyed by blueprint slug.
 */
export async function applyBlueprints(
  slugs: string[],
  params?: Record<string, Record<string, string>>,
  source?: string,
  authToken?: string,
): Promise<BlueprintApplyResult[]> {
  return fetchApi<BlueprintApplyResult[]>(
    `/api/groups/blueprints/apply${query({ source })}`,
    { method: "POST", body: JSON.stringify({ slugs, params: params ?? null }), headers: { "Content-Type": "application/json" } },
    authToken,
  );
}
